Fix slider button tracks and general replay accuracy

peppy opened the pull request:
Replaces #4065.

Fixes https://github.com/ppy/osu/issues/2647
- [x] Depends on https://github.com/ppy/osu-framework/pull/2166

peppy committed:
Slider press fix

nyquillerium committed:
Add tests for slider input behavior

peppy committed:
Add fixes

peppy committed:
Add extensive commenting

ekrctb commented on osu.Game.Rulesets.Osu.Tests/TestCaseSliderInput.cs:
```diff
@@ -0,0 +1,406 @@
+        ///     should result in tracking to end.
+        ///     At 250ms intervals:
+        ///     Frame 1 (prior to slider):          Left Click
+        ///     Frame 2 (within slider hit window): Left & Right Click
```
& -> &amp; (same below)

nyquillerium committed:
Update test documentation

peppy committed:
Add comment about other cases, rewind handling

nyquillerium committed:
Change ampersands in xmldoc

smoogipoo commented on osu.Game.Rulesets.Osu/Objects/Drawables/Pieces/SliderBall.cs:
```diff
@@ -145,22 +148,77 @@ private set
+        ///
+        /// In all other cases, no special handling is required (either key being pressed is allowable as valid tracking).
+        ///
+        /// The reason for storing this as a time value (rather than a bool) is to correct handle rewind scenarios.
```
```suggestion
        /// The reason for storing this as a time value (rather than a bool) is to correctly handle rewind scenarios.
```

smoogipoo committed:
Fix typo in comment

Co-Authored-By: peppy <pe@ppy.sh>

peppy commented on osu.Game.Rulesets.Osu.Tests/TestCaseSliderInput.cs:
```diff
@@ -0,0 +1,410 @@
+        }
+
+        /// <summary>
+        ///     Pressing a key before a slider, hitting a slider head, leaving the slider, then coming back into the slider to track it should NOT start retracking
```
wrong documentation. need to mention hitting with a second key and also releasing the key?

nyquillerium replied:
The releasing key test is test 9, but yeah need to mention the hitting second key part

peppy commented on osu.Game.Rulesets.Osu.Tests/TestCaseSliderInput.cs:
```diff
@@ -0,0 +1,410 @@
+        }
+
+        /// <summary>
+        ///     Halfway into a slider outside of the slider, then starting to hover the slider afterwards should result in tracking
```
i think this is broken english "Halfway into a slider outside of the slider"

nyquillerium committed:
Reword slider test xmldocs.

peppy committed:
Formatting and constants

peppy committed:
Use schedule for *all* setup

peppy committed:
Remove unnecessary content override

peppy committed:
Fix dynamic compilation failing

nyquillerium committed:
Split out testcase player for use in slider input tests

nyquillerium committed:
Fix auto-generated lisence header

nyquillerium committed:
Re-order tests

nyquillerium committed:
Fix input tests not using async loads

nyquillerium committed:
re-order tests again

nyquillerium committed:
Check for beatmap load state before performing test

nyquillerium committed:
Fix dynamic compilation, use slidertick judgements

nyquillerium committed:
Nuke whitespace

nyquillerium committed:
phantom whitespace

nyquillerium committed:
Nuke whitespace again

nyquillerium committed:
Move all beatmap initialization logic into constructor

nyquillerium committed:
Remove load checks and place one in the constructor

nyquillerium committed:
Rename base player test class to RateAdjustedBeatmap, refactor input tests

peppy committed:
Update headers

peppy committed:
Make base class abstract and add documentation

peppy committed:
Formatting, ordering and simplification

peppy committed:
Not sure what a bm is

peppy committed:
Fix some huge oversights

peppy committed:
Avoid test frame being the precise time of slider end

peppy committed:
Ensure variable isolation over multiple test runs

peppy committed:
Fix tests not running more than once

peppy committed:
Use game clock as reference

peppy commented:
Have run each test over 2,000 times with zero failures.

peppy committed:
Apply new RulesetInputManager logic

Run UpdateSubTree twice to ensure correctness

peppy committed:
Move beatmap construction to step

peppy committed:
Simplify SliderBall and fix incorrect key up handling

Was not processing timeToAcceptAnyKeyAfter when cursor was outside valid tracking area, but should have been.

peppy committed:
Remove unnecessary using

peppy committed:
Holy sheet

peppy committed:
Simplify RulesetInputManager further

peppy committed:
Update framework

peppy commented:
This is finally ready for review.

smoogipoo committed:
Cleanup

smoogipoo committed:
Rename testcase

peppy committed:
Merge branch 'master' into slider-fix
